Add summary pings

marcrowo opened the pull request:
~~Based on #33 `add-variations`.~~

marcrowo committed:
Update min Firefox version to 55 in install.rdf

marcrowo committed:
Added sending of summary ping.

This summary ping will also conform to the parquet schema enforced
by the studyUtils telemetry() function.

marcrowo committed:
Upgrade CircleCI environment and npm scripts.

marcrowo committed:
Updgraded Docker image version number.

marcrowo committed:
Update Docker image again to fix CircleCI.

Problem: Docker working directory and git checkout directory
were the same.

marcrowo committed:
Fixed tar extracting to CircleCI working_directory.

marcrowo committed:
Actually upgrade Docker image.

marcrowo committed:
Install firefox for dependencies.

marcrowo committed:
Copy tests debugging.

marcrowo committed:
More copy testing.

marcrowo committed:
Switch copy controller order.

marcrowo committed:
Revert debugging changes to button_test.js.

marcrowo committed:
Updated shell script for Docker setup.

marcrowo committed:
Removed git checkout from setup script.

marcrowo committed:
Made Docker script executable.

marcrowo committed:
Run commands in the background in setup.

marcrowo committed:
Updated setup script comment for . ./docker_setup.

marcrowo committed:
Increase timeout for highlight treatment tests.

marcrowo committed:
More timeout increases.

marcrowo committed:
More timeout.

marcrowo committed:
Use waitForClassAdded() again.

marcrowo committed:
Added try-catch to waitForClassAdded().

marcrowo committed:
Fixed linting issue.

marcrowo committed:
Removed comment.

marcrowo committed:
Basic summary ping test.

marcrowo committed:
Refactored test suite.

- animation / popup test
- telemetry searching
- overflow test

marcrowo committed:
Added summary ping tests.

- made telemetry() function return promise
- updated eslint

marcrowo committed:
Fixed linting issues in bootstrap.js

marcrowo committed:
Increase timeout for summary ping tests.

marcrowo committed:
Removed summary ping test from highlight tests.

marcrowo committed:
Linting fix.

marcrowo committed:
Added timeout for hasShareButton false test.

marcrowo committed:
Added try-catch around waitForAnimationEnd().

marcrowo committed:
Fixed linting issue.

marcrowo committed:
Increased timeout for clicks.

marcrowo committed:
Better panel closing to fix tests.

marcrowo commented:
Updated test list below. These tests are still not always consistent.

```
Basic Functional Tests
  - should have a URL bar
  - should have a share button
  - should have copy paste working
  Addon uninstall tests
    - should no longer trigger animation once uninstalled
    - should no longer trigger popup once uninstalled
    - should no longer trigger ask panel once uninstalled
    - should not add the button to the toolbar once uninstalled
Highlight Treatment Tests
  - animation should trigger on regular page
  - animation should not trigger on disabled page
  - animation should not trigger if the share button is not added to toolbar
  - should not trigger animation if the share button is in the overflow menu
  - should send highlight and copy telemetry pings
Summary Ping Tests
  - should set hasShareButton to false if the share button is not added
  - should set hasShareButton to true if the share button is added
  - should report the correct number of URL copy events
  - should report the correct number of share button clicks
  - should report the correct number of share panel clicks
  - should log a summary ping for highlight treatment
  - should log a summary ping for doorhangerDoNothing treatment
  - should log a summary ping for doorhangerAskToAdd treatment
  - should log a summary ping for doorhangerAddToToolbar treatment
DoorhangerDoNothing Treatment Tests
  - popup should trigger on regular page
  - popup should not trigger on disabled page
  - popup should not trigger if the share button is not added to toolbar
  - should not trigger doorhanger if the share button is in the overflow menu
  - should send doorhanger and copy telemetry pings
DoorhangerAskToAdd Treatment Tests
  - should open an ask panel on a regular page without the share button
  - should open a standard panel on a regular page with the share button
  - should not open an ask panel on a regular page with the share button
  - should not open an ask panel on a regular page if the share button is in the overflow menu
  - should not open an ask panel on a disabled page
  - should add the button to the toolbar upon clicking on ask panel
DoorhangerAddToToolbar Treatment Tests
  - should add the button to the toolbar upon copy paste on regular page
  - popup should trigger on regular page
  - animation should trigger on regular page
```

mythmon commented on extension/PingStorage.jsm:
```diff
@@ -0,0 +1,67 @@
+
+  async logPing(pingData) {
+    const db = await getDatabase();
+    const ping = Object.assign({ timestamp: new Date().toISOString() }, pingData);
```
You can store dates directly in indexeddb, so it would probably be better to avoid the serialization here. It's best to store the richest data you have, so that you have maximum flexibility later.

marcrowo replied:
Note: the original intention behind this decision was to match the Telemetry ping timestamp format.

mythmon commented on extension/bootstrap.js:
```diff
@@ -163,43 +173,48 @@ class CopyController {
-        curController.doCommand(cmd);
-        break;
+        // FIXME Cannot await within CopyController
+        await TREATMENTS[this.treatment](this.browserWindow, shareButton);
```
You say "cannot await", and then immediately await? I'm not sure I understand this comment.

marcrowo replied:
I forgot to remove the `FIXME` after I fixed the issue (namely, that the CopyController was not `async` because I was not sure this would work).

mythmon commented on test/button_test.js:
```diff
@@ -11,33 +11,103 @@ const webdriver = require("selenium-webdriver");
+    const callback = args[args.length - 1];
+    Components.utils.import("resource://gre/modules/Preferences.jsm");
+    // using the rest parameters, treatment = args[0]
+    Preferences.set("extensions.sharebuttonstudy.treatment", args[0]);
```
You can probably convert this to explicit arguments instead of `...args`.

marcrowo committed:
Implemented local storage of pings in IndexedDB.

Notable changes includes the new PingStorage class and upgrading to
locally to Firefox Beta for v55.

marcrowo committed:
Changed logPing() to use Date object since IndexDB supports this.

marcrowo committed:
Removed extraneous FIXME's.

marcrowo committed:
Replaced spread operator ...args with explicit arguments.
